ui/src/reusable_ui/components/dropdowns/Dropdown.tsx

```diff
@@ -0,0 +1,174 @@
+  }
+
+  constructor(props: Props) {
+    super(props)
+
+    this.state = {
+      expanded: false,
+    }
+  }
+
+  public render() {
+    const width = `${this.props.width}px`
+
+    return (
+      <ClickOutside onClickOutside={this.collapseMenu}>
+        <div className={this.containerClassName} style={{width}}>
+          {this.button}
+          {this.menu}
+        </div>
+      </ClickOutside>
+    )
+  }
+
+  private toggleMenu = (): void => {
+    this.setState({expanded: !this.state.expanded})
+  }
+
+  private collapseMenu = (): void => {
+    this.setState({expanded: false})
+  }
+
+  private get containerClassName(): string {
+    const {color, size, disabled, wrapText, customClass} = this.props
+
+    return classnames(`dropdown dropdown-${size} dropdown-${color}`, {
+      disabled,
+      'dropdown-wrap': wrapText,
+      [customClass]: customClass,
+    })
+  }
+
+  private get button(): JSX.Element {
+    const {selectedItem, disabled, color, size, icon} = this.props
+    const {expanded} = this.state
+
+    return (
+      <DropdownButton
+        label={selectedItem}
+        active={expanded}
+        color={color}
+        size={size}
+        icon={icon}
+        disabled={disabled}
+        onClick={this.toggleMenu}
+      />
+    )
+  }
+
+  private get menu(): JSX.Element {
+    const {selectedItem, maxMenuHeight, menuColor} = this.props
+    const {expanded} = this.state
+
+    if (expanded) {
+      return (
+        <div
+          className={`dropdown--menu-container dropdown--${menuColor}`}
+          style={this.menuStyle}
+        >
+          <FancyScrollbar
+            autoHide={false}
+            autoHeight={true}
+            maxHeight={maxMenuHeight}
+          >
+            <div className="dropdown--menu">
+              {this.flatChildren.map((child: JSX.Element) =>
+                React.cloneElement(child, {
+                  ...child.props,
+                  key: uuid.v4(),
+                  selected: child.props.text === selectedItem,
+                  onClick: this.handleItemClick,
+                })
+              )}
+            </div>
+          </FancyScrollbar>
+        </div>
+      )
+    }
+
+    return null
+  }
+
+  private get flatChildren() {
+    const children = React.Children.toArray(this.props.children)
+
+    const childrenWithoutFragments = children.map((child: JSX.Element) => {
+      if (child.type === Fragment) {
+        const childArray = React.Children.toArray(child.props.children)
+        return childArray
+      }
+
+      return child
+    })
+
+    return _.flattenDeep(childrenWithoutFragments)
+  }
+
+  private get menuStyle(): CSSProperties {
+    const {wrapText, width} = this.props
+
+    if (wrapText) {
+      return {
+        width: `${width}px`,
+      }
+    }
+
+    return {
+      minWidth: `${width}px`,
+    }
+  }
+
+  private handleItemClick = (value: any) => {
+    const {onChange} = this.props
+    onChange(value)
```

Comment:
what happens if whatever this `onChange` maps to, fails at any time? would you still want to collapse the menu? is it possible that the `onChange` ever maps to an async function and you may want to `await` and surround with a `try/catch`?

Reply:
It is typed as a non-async function in props should it should prevent misuse